### PR TITLE
bt-326: Save badges and hard skills for user details

### LINKED ISSUE #326: Connect badges & hard skills saving for user details
Save badges & hard skills from Onboarding (Screenshot 1 & Screenshot 2)

(Screenshot 1)
Image

(Screenshot 2)
Image

### PR DESCRIPTION
**Pull request by Simpsonkk**

### CODE REVIEW

File: backend/src/migrations/20230827061743_add_talent_badges_table.ts
Comment (nika-sirak) [inline backend/src/migrations/20230827061743_add_talent_badges_table.ts:1]: We are not supposed to make any manipulations with already created migrations that have been merged.

File: backend/src/migrations/20230828083409_update_users_table.ts
Comment (nika-sirak) [inline backend/src/migrations/20230828083409_update_users_table.ts:1]: The same

**Review by nika-sirak — CHANGES_REQUESTED** [review]

File: backend/src/migrations/20230827061743_add_talent_badges_table.ts
Comment (Simpsonkk) [reply backend/src/migrations/20230827061743_add_talent_badges_table.ts:1]: I reverted the old data and updated the talent badge table using migrate:dev:make

File: backend/src/migrations/20230828083409_update_users_table.ts
Comment (Simpsonkk) [reply backend/src/migrations/20230828083409_update_users_table.ts:1]: I reverted the old data and updated the talent badge table using migrate:dev:make

File: backend/src/bundles/user-details/user-details.service.ts [hunk 1]
```diff
@@ -31,14 +37,27 @@ class UserDetailsService implements Service {
...
         payload: UserDetailsCreateRequestDto,
     ): Promise<UserDetailsResponseDto> {
-        const newUserDetails = await this.userDetailsRepository.create(payload);
+        const { talentBadges, ...userDetails } = payload;
+
+        const newUserDetails = await this.userDetailsRepository.create(
+            userDetails,
+        );
+
+        talentBadges?.map(async (talentBadge) => {
```
Comment (Monyster) [inline backend/src/bundles/user-details/user-details.service.ts]: Promise.all(...)

Comment (Simpsonkk) [reply backend/src/bundles/user-details/user-details.service.ts]: added

File: backend/src/bundles/talent-badges/talent-badge.repository.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    public async update({
+        id,
+        isShown,
+        userDetailsId,
+    }: TalentBadgePatchAndFetch): Promise<TalentBadgeEntity> {
+        const talentBadge = await this.talentBadgeModel
+            .query()
+            .patchAndFetchById(id, {
+                isShown: !isShown,
```
Comment (Monyster) [inline backend/src/bundles/talent-badges/talent-badge.repository.ts]:
```suggestion
                isShown,
```
Why did you reverse it?

File: backend/src/bundles/talent-badges/talent-badge.service.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    public findAll(): Promise<{ items: TalentBadgeRepository[] }> {
+        throw new Error(ErrorMessages.NOT_IMPLEMENTED);
+    }
+
+    public find(): Promise<TalentBadgeRepository> {
+        throw new Error(ErrorMessages.NOT_IMPLEMENTED);
+    }
+
+    public async create(badge: TalentBadgeCreate): Promise<TalentBadgeEntity> {
+        return await this.talentBadgeRepository.create(badge);
```
Comment (Monyster) [inline backend/src/bundles/talent-badges/talent-badge.service.ts]: I see a lot same return statements with await.
You have already return Promise, you do not need to await it
Read more [no-return-await](https://kanby.medium.com/%D0%BF%D0%BE%D1%87%D0%B5%D0%BC%D1%83-%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D1%82%D1%8C-return-await-%D0%BF%D0%BB%D0%BE%D1%85%D0%B0%D1%8F-%D0%B8%D0%B4%D0%B5%D1%8F-e87b70015f0c)
```suggestion
        return this.talentBadgeRepository.create(badge);
```

File: backend/src/bundles/talent-badges/types/talent-badge-update.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    id: string;
+    userId: string;
+    badgeId: string;
+    isShown?: boolean;
+    userDetailsId: string | null;
+};
+
+type TalentBadgeUpdate = Omit<TalentBadgePatchAndFetch, 'id'>;
+
+export { type TalentBadgePatchAndFetch, type TalentBadgeUpdate };
```
Comment (Monyster) [inline backend/src/bundles/talent-badges/types/talent-badge-update.ts]: Filename is "...-update.ts" but you returning 2 different types.

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import { type Entity } from '~/common/types/types.js';
+
+class TalentHardSkillsEntity implements Entity {
+    private 'id': string | null;
+
+    private 'hardSkillId': string;
+
+    private 'userDetailsId': string | null;
```
Comment (Monyster) [inline backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts]: Hard-skill assigned to user-details can be null?
```suggestion
    private 'userDetailsId': string;
```

File: shared/src/bundles/user-details/types/user-details-create-dto.type.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    description?: string;
+
+    englishLevel?: ValueOf<typeof EnglishLevel>;
+    notConsidered?: ValueOf<typeof NotConsidered>[];
+    preferredLanguages?: ValueOf<typeof PreferredLanguages>[];
+
+    projectLinks?: string[];
+    photoId?: string;
+    fullName: string;
```
Comment (Monyster) [inline shared/src/bundles/user-details/types/user-details-create-dto.type.ts]: We can make it optional (As improvement)
Example
```js
            const newUserDetails = await this.userDetailsRepository.create({
                fullName: fullName ?? '',
                ...payload,
            });
```

File: shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts [hunk 7]
```diff
@@ -1,49 +1,12 @@
...
-    photoId?: string;
-    fullName?: string;
-    phone?: string;
-    linkedinLink?: string;
-    companyName?: string;
-    companyLogoId?: string;
-    companyWebsite?: string;
-    employerPosition?: string;
-    cvId?: string;
+    talentBadges?: string[];
```
Comment (Monyster) [inline shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts:7]: I guess that's can be as below example or not?
```
    talentBadges?: {
        userId: string;
        score: number;
        level: string;
        badgeId: string;
    }[];
```

File: shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts [hunk 8]
```diff
@@ -64,6 +64,8 @@ const userDetailsCreate = joi.object<UserDetailsCreateRequestDto, true>({
     companyWebsite: joi.string().trim().uri(),
     employerPosition: joi.string().trim(),
     cvId: joi.string().trim(),
+    talentBadges: joi.array().items(joi.string().trim()),
```
Comment (Monyster) [inline shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts:67]: talentBadges: joi
        .array()
        .items(
            joi.object({
                userId: joi.string().required(),
                score: joi.number().required(),
                level: joi.string().required(),
                badgeId: joi.string().required(),
            }),
        )

File: shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts [hunk 9]
```diff
@@ -68,6 +68,8 @@ const userDetailsUpdate = joi
         companyWebsite: joi.string().trim().uri(),
         employerPosition: joi.string().trim(),
         cvId: joi.string().trim(),
+        talentBadges: joi.array().items(joi.string().trim()),
```
Comment (Monyster) [inline shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts:71]: same

**Review by Monyster — CHANGES_REQUESTED** [review]
Don't forget swagger docs

File: backend/src/bundles/talent-badges/talent-badge.repository.ts (hunk 2 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: In the second step we have only checkboxes and as I understand we should just reverse isShown of this badge if we already have it for the user in DB.

Comment (Monyster) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: Backend part should make sense, and that logic should be in the frontend part.

Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: as I undersood @MaksymGavatuk about talent badges, we should recieve from frontend only array of strings = Ids also without isShown and other fields

File: backend/src/bundles/talent-badges/talent-badge.service.ts (hunk 3 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.service.ts]: thanks

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts (hunk 5 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts]: according to our migration add_talent_hard_skills_table.ts user-details can be null

File: shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts (hunk 7 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts:7]: as I understood @MaksymGavatuk about talent badges, we should recieve from frontend only array of strings = Ids

File: shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts (hunk 8 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts:67]: as I understood @MaksymGavatuk about talent badges, we should recieve from frontend only array of strings = Ids

File: shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts (hunk 9 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts:71]: as I understood @MaksymGavatuk about talent badges, we should recieve from frontend only array of strings = Ids

File: backend/src/bundles/talent-badges/types/talent-badge-update.ts (hunk 4 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/types/talent-badge-update.ts]: divided

File: backend/src/bundles/talent-badges/talent-badge.repository.ts (hunk 2 above)
Comment (MaksymGavatuk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: Yes it just string[]

File: shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts (hunk 7 above)
Comment (MaksymGavatuk) [reply shared/src/bundles/user-details/types/user-details-update-request-dto.type.ts:7]: Yes it just string[]

File: shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts (hunk 8 above)
Comment (MaksymGavatuk) [reply shared/src/bundles/user-details/validation-schemas/user-details-create.validation-schema.ts:67]: Yes it just string[]

File: shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts (hunk 9 above)
Comment (MaksymGavatuk) [reply shared/src/bundles/user-details/validation-schemas/user-details-update.validation-schema.ts:71]: Yes it just string[]

File: backend/src/bundles/talent-badges/talent-badge.repository.ts (hunk 2 above)
Comment (MaksymGavatuk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: Yes, but in this case, we need to set isShow as we provide from params

File: backend/src/bundles/talent-badges/talent-badge.repository.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    public async update({
+        id,
+        isShown,
+        userDetailsId,
+    }: TalentBadgePatchAndFetch): Promise<TalentBadge> {
+        const talentBadge = await this.talentBadgeModel
+            .query()
+            .patchAndFetchById(id as string, {
+                isShown: !isShown,
```
Comment (MaksymGavatuk) [inline backend/src/bundles/talent-badges/talent-badge.repository.ts]:
```suggestion
                isShown,
```

File: shared/src/bundles/user-details/types/user-details-response-dto.type.ts [hunk 11]
```diff
@@ -37,6 +37,20 @@ type UserDetailsResponseDto = {
     companyWebsite: string | null;
     employerPosition: string | null;
     cvId: string | null;
+    talentBadges?: {
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/types/user-details-response-dto.type.ts]: Why optional, just empty array

File: shared/src/bundles/user-details/types/user-details-response-dto.type.ts [hunk 12]
```diff
@@ -37,6 +37,20 @@ type UserDetailsResponseDto = {
...
+        id: string | null;
+        userId: string;
+        score: number | null;
+        level: string | null;
+        badgeId: string;
+        isShown: boolean;
+        userDetailsId: string | null;
+    }[];
+    talentHardSkills?: {
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/types/user-details-response-dto.type.ts]: Why optional, just empty array

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts (hunk 5 above)
Comment (MaksymGavatuk) [reply backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts]: Can you please explain why userDetailsId can be null ?

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.repository.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            await this.talentHardSkillsModel
+                .query()
+                .insert(skillsToInsert)
+                .execute();
+        }
+
+        return this.findByUserDetailsId(userDetailsId);
+    }
+
+    public async deleteUnusedHardSkills(
```
Comment (MaksymGavatuk) [inline backend/src/bundles/talent-hard-skills/talent-hard-skills.repository.ts]:
```suggestion
    public async deleteByIds(
```

File: backend/src/bundles/talent-badges/talent-badge.service.ts [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        const badge = await this.talentBadgeRepository.find({
+            badgeId,
+            userId,
+        });
+
+        return badge
+            ? this.talentBadgeRepository.update({
+                  ...badge,
+                  userDetailsId,
+              })
```
Comment (MaksymGavatuk) [inline backend/src/bundles/talent-badges/talent-badge.service.ts]: We need to provide isShown option here

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: backend/src/bundles/talent-badges/talent-badge.repository.ts (hunk 2 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: done

File: shared/src/bundles/user-details/types/user-details-create-dto.type.ts (hunk 6 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/types/user-details-create-dto.type.ts]: added

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts (hunk 5 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-hard-skills/talent-hard-skills.entity.ts]: you are right. I removed null

File: backend/src/bundles/talent-badges/talent-badge.repository.ts (hunk 10 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.repository.ts]: done

File: shared/src/bundles/user-details/types/user-details-response-dto.type.ts (hunk 11 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/types/user-details-response-dto.type.ts]: done

File: shared/src/bundles/user-details/types/user-details-response-dto.type.ts (hunk 12 above)
Comment (Simpsonkk) [reply shared/src/bundles/user-details/types/user-details-response-dto.type.ts]: done

File: backend/src/bundles/talent-hard-skills/talent-hard-skills.repository.ts (hunk 13 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-hard-skills/talent-hard-skills.repository.ts]: done

File: backend/src/bundles/talent-badges/talent-badge.service.ts (hunk 14 above)
Comment (Simpsonkk) [reply backend/src/bundles/talent-badges/talent-badge.service.ts]: done

File: mobile/src/bundles/talent/store/actions.ts [hunk 15]
```diff
@@ -36,7 +36,8 @@ const updateOnboardingData = createAsyncThunk<
     const { badges, hardSkills, photo, cv, ...payload } = stepPayload;
 
     if (Object.keys(payload).length === 0) {
-        return stepPayload as UserDetailsGeneralResponseDto;
+        //TODO remove unknown
+        return stepPayload as unknown as UserDetailsGeneralResponseDto;
```
Comment (MaksymGavatuk) [inline mobile/src/bundles/talent/store/actions.ts:40]: TODO ?

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: mobile/src/bundles/talent/store/actions.ts (hunk 15 above)
Comment (Simpsonkk) [reply mobile/src/bundles/talent/store/actions.ts:40]: @MaksymGavatuk yep, I wrote this TODO for mobile team and I told them about it

**Review by MaksymGavatuk — APPROVED** [review]

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
should we update the frontend here? 
if no, do we have a ticket for the frontend?

**Comment by Simpsonkk** [comment]
> should we update the frontend here?
> if no, do we have a ticket for the frontend?

if you about mobile changes, I did changes in it because of linter errors from mobile folder. I communicated with mobile team they knew it.

**Pull request merged**